tx/group compaction fixes

bharathv opened the pull request:
Check individual commit message for details

Main changes

* Adds a lots of observability
* * Now consumer offsets partitions are supported in describe_producers Kafka API
* * Adds a v1/producers/kafka/topic/partition end point that gives detailed producer level debug info
* Adds an escape hatch to force abort any stuck group transactions  `/v1/transaction/unsafe_abort_group_transaction/{group_id}`
* Fixes a couple of compaction bugs with group transactions

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.3.x
- [x] v24.2.x
- [ ] v24.1.x

## Release Notes
### Bug Fixes
* Fixes an issue that blocked the compaction of consumer offsets with group transactions.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
tx/group: track begin offset of transactions

bharathv committed:
tx/tests/dt: test for describe producers

bharathv committed:
tx/groups: escape hatch for unsafe aborting of group transactions

This is unsafe because it does not do any required checks to see
if a particular transaction is in progress and is a candidate for abort.
For example if a transaction is committed by the coordinator and
pending commit on the group, using this escape hatch to abort the
transaction can cause correctness issues. To be used with caution as an
escape hatch for aborting transactions that the group has lost track of
are ok to be aborted. This situation usually is indicative of a bug in
the transaction implementation.

bharathv committed:
group/tx: add a ducktape test for compactibility of consumer_offsets

bharathv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f612a4c1-7d4d-4480-8d27-b82cf1e21b66-->
<!--TEST RESULTS-->
<!--3dabf01a-895b-41eb-a431-22e5f8301ab3-->
<!--TEST RESULTS-->
<!--df132139-b7de-40c8-9832-11ffa55679d8-->
<!--TEST RESULTS-->
<!--38e645c9-1240-4ecb-8ef7-831f41ca5cb2-->
<!--TEST RESULTS-->
<!--71b09f33-ad76-4607-a1e9-fe7b48258122-->
<!--TEST RESULTS-->
<!--3321bf8e-ffe6-4406-915f-0f804e75ee89-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60037">build#60037</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| test_consumer_group_recovery_rpunit.test_consumer_group_recovery_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60037#0193e7ef-114e-474c-b463-9426fd2ed955 | FAIL | 0/2 |
| test_consumer_group_recovery_rpunit.test_consumer_group_recovery_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60037#0193e7ef-1152-4aea-a1e2-19d14ea9a00a | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60049">build#60049</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/60049#0193f07e-533f-4715-b8a1-0f31c067b1da | FLAKY | 1/2 |
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0bf-dcc3-4557-af8c-d9940a72aff1 | FAIL | 0/1 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0bf-dcc1-4950-bb16-fd655ecc86af | FLAKY | 3/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0bf-dcc2-4f4b-8cde-76741d798378 | FLAKY | 4/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0c5-171e-474d-aa8e-0d0ffb161139 | FLAKY | 1/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0c5-171f-4e2a-8beb-403808ca77b4 | FLAKY | 2/6 |
| rptest.tests.e2e_topic_recovery_test.EndToEndTopicRecovery.test_restore_with_aborted_tx.recovery_overrides=.retention.local.target.bytes.1024.redpanda.remote.write.True.redpanda.remote.read.True.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0c5-171b-40a1-bfbf-37f5a6e814bf | FAIL | 0/1 |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60049#0193f0c5-1719-46be-8112-309c6520cae5 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60058">build#60058</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_storage_scrubber_test.CloudStorageScrubberTest.test_scrubber.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60058#0193f31d-2f6d-4a69-8dc9-12c8e28584c6 | FAIL | 0/1 |
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_mount_inexistent | ducktape | https://buildkite.com/redpanda/redpanda/builds/60058#0193f337-48f5-4f5d-b1f2-e3b95d1ca948 | FLAKY | 5/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60058#0193f31d-2f6c-47ff-9829-30ea21edb404 | FLAKY | 2/6 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/60058#0193f31d-2f6f-44e5-8895-92629bdc07fb | FAIL | 0/1 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/60058#0193f337-48f5-4f5d-b1f2-e3b95d1ca948 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60103">build#60103</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60103#0193f706-0cb5-4c9d-88ba-d8b9ef93524e | FLAKY | 2/6 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/60103#0193f702-3285-483a-b9fb-81658e3e80bb | FAIL | 0/1 |
| rptest.transactions.producers_api_test.ProducersAdminAPITest.test_producers_state_api_during_load | ducktape | https://buildkite.com/redpanda/redpanda/builds/60103#0193f706-0cb5-4c9d-88ba-d8b9ef93524e | FAIL | 0/1 |
| test_group_tx_compaction_rpunit.test_group_tx_compaction_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60103#0193f6bf-f184-47e1-84a0-4b121a138781 | FAIL | 0/2 |
| test_group_tx_compaction_rpunit.test_group_tx_compaction_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60103#0193f6bf-f184-4e44-8109-e4d183911389 | FAIL | 0/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60115">build#60115</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| partition_balancer_simulator_test_rpunit.partition_balancer_simulator_test_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/60115#0193f7b9-a957-48fc-8c6a-db1d4996766d | FLAKY | 1/2 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60115#0193f800-181d-4b4b-a17b-030b181e6427 | FLAKY | 3/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/60115#0193f800-181e-4a88-93a8-db3096d09b07 | FLAKY | 4/6 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/60115#0193f812-e6f8-41cc-8c78-144da0e50294 | FLAKY | 5/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60249">build#60249</a></summary>


| test_id                                                                                                                       | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.data_migrations_api_test.DataMigrationsApiTest.test_mount_inexistent                                             | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60249#01942ae9-8ee8-4ffe-ae4e-71370ce7f503 | FLAKY         | 5/6      |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60249#01942ae9-8ee8-4ffe-ae4e-71370ce7f503 | FAIL          | 0/1      |
| rptest.transactions.consumer_offsets_test.VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets.versions_to_upgrade=3 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60249#01942aed-863f-46ee-8a8a-c6505749bd17 | FAIL          | 0/1      |
</details>
<!--/body-->

bharathv committed:
tx/observability: add types and plumbing needed to get producer states

.. for a given partition, to be hooked up with REST API in the next
commit.

bharathv committed:
tx/producer_state: add getters for internal state

bharathv committed:
tx/admin: types for exposing producer info in REST api

bharathv committed:
tx/observability: REST endpoint to fetch all producers from a partition

/v1/debug/producers/{namespace}/{topic}/{partition}

.. includes low level debug information about producers for
idempotency/transactional state.

bharathv commented:
/ci-repeat 3

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60049-->
<!--f033946e-b0b7-4f8a-9b75-266b4376646f-->
<!--b1b4f2b1-4534-494b-9345-9ce77596abe9-->
<!--d04cac75-8cf8-49c7-baba-237d3c002682-->
<!--title-->
## Retry command for [Build#60049](https://buildkite.com/redpanda/redpanda/builds/60049)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":1}
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":3}
tests/rptest/tests/e2e_topic_recovery_test.py::EndToEndTopicRecovery.test_restore_with_aborted_tx@{"cloud_storage_type":2,"recovery_overrides":{"redpanda.remote.read":true,"redpanda.remote.write":true,"retention.local.target.bytes":1024}}
```
<!--/body-->

bharathv committed:
k/group: disallow group deletion while transactions in progress

This will result in hanging transactions and subsequent blocking
of compaction.

mmaslankaprv commented on src/v/kafka/server/group_manager.h:
```diff
@@ -185,6 +186,8 @@ class group_manager {
 
     described_group describe_group(const model::ntp&, const kafka::group_id&);
 
+    partition_response describe_partition_producers(const model::ntp&);
```
nit: `partition_response` is confusing as it has no context, can we create an alias for this type f.e. `partition_producers` ?

bharathv replied:
sure, `partition_response` is kafka API parlance, will typedef.

mmaslankaprv commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1620,6 +1621,64 @@ group_manager::describe_group(const model::ntp& ntp, const kafka::group_id& g) {
+    }
+    response.error_code = kafka::error_code::none;
+    // snapshot the list of groups attached to this partition
+    fragmented_vector<std::pair<group_id, group_ptr>> groups;
```
chunked_vector ?

mmaslankaprv commented on src/v/kafka/server/group_tx_tracker_stm.h:
```diff
@@ -72,31 +96,9 @@ class group_tx_tracker_stm final
-
-        absl::btree_map<model::producer_identity, model::offset>
-          producer_to_begin;
+    const all_txs_t& inflight_transactions() const { return _all_txs; }
```
This part seems not relevant for current commit ?

bharathv replied:
done.

mmaslankaprv commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -52,12 +52,30 @@ ss::future<> group_tx_tracker_stm::do_apply(const model::record_batch& b) {
+                for (auto& [pid, offset] : group_state.producer_to_begin) {
+                    vlog(
+                      klog.trace,
+                      "[{}] group: {}, producer: {}, begin: {}",
+                      _raft->ntp(),
```
nit: can we add more context to this log line  ? i.e. it is not known what the begin is

bharathv replied:
removed this, don't think this is needed.

mmaslankaprv commented on src/v/redpanda/admin/transaction.cc:
```diff
@@ -213,3 +221,73 @@ admin_server::delete_partition_handler(std::unique_ptr<ss::http::request> req) {
+
+    if (pid_str.empty() || epoch_str.empty() || sequence_str.empty()) {
+        throw ss::httpd::bad_param_exception(
+          "invalid producer_id/epoch, should be >= 0");
```
nit: the error message is misaligned

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60058-->
<!--2a4ce9b4-3f7c-42c3-af71-122296e09c94-->
<!--e3fda9fa-36b2-49db-8ae1-9ab87c406016-->
<!--b544ebbd-1a91-40ac-9f7d-1178ec9045a2-->
<!--title-->
## Retry command for [Build#60058](https://buildkite.com/redpanda/redpanda/builds/60058)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```


/ci-repeat 1
tests/rptest/transactions/producers_api_test.py::ProducersAdminAPITest.test_producers_state_api_during_load
tests/rptest/tests/cloud_storage_scrubber_test.py::CloudStorageScrubberTest.test_scrubber@{"cloud_storage_type":1}
```
<!--/body-->

nvartolomei commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1620,6 +1621,64 @@ group_manager::describe_group(const model::ntp& ntp, const kafka::group_id& g) {
+            auto& tx = state.transaction;
+            int64_t start_offset = -1;
+            if (tx && tx->begin_offset >= model::offset{0}) {
+                start_offset = partition->get_offset_translator_state()
```
is it possible that `tx->begin_offset` doesn't exist anymore/removed by retention?

bharathv replied:
no its not, we don't apply delete retention on consumer_offsets and even if we do we GC producers whose last offset falls below the retained offset.

nvartolomei commented on src/v/cluster/tx_gateway_frontend.cc:
```diff
@@ -2974,4 +2974,23 @@ ss::future<tx::errc> tx_gateway_frontend::do_delete_partition_from_tx(
     co_return tx::errc::none;
 }
 
+ss::future<tx::errc> tx_gateway_frontend::unsafe_abort_group_transaction(
```
can you add a comment describing what is "unsafe" about this operation? what invariants will break? what semantic behavior breaks?

bharathv replied:
done.

nvartolomei commented on src/v/redpanda/admin/api-doc/transaction.json:
```diff
@@ -92,6 +92,46 @@
+                        },
+                        {
+                            "name": "sequence",
+                            "in": "int",
```
typo? should be `"in": "query",`

bharathv replied:
done.

nvartolomei commented on src/v/redpanda/admin/transaction.cc:
```diff
@@ -213,3 +221,73 @@ admin_server::delete_partition_handler(std::unique_ptr<ss::http::request> req) {
+    }
+
+    if (pid_str.empty() || epoch_str.empty() || sequence_str.empty()) {
+        throw ss::httpd::bad_param_exception(
```
This should happen out of the box btw. This seems to be called implicitly when handling requests https://github.com/redpanda-data/seastar/blob/09a59a23ff2740a2fa591b0e65d978ca83d2b9e3/include/seastar/http/handlers.hh#L76

bharathv replied:
ya pretty conservative check, was thinking to convert to an assert but probably not worth it.

nvartolomei commented on src/v/redpanda/admin/transaction.cc:
```diff
@@ -213,3 +221,73 @@ admin_server::delete_partition_handler(std::unique_ptr<ss::http::request> req) {
+    auto sequence_str = request->get_query_param("sequence");
+
+    if (group_id.empty()) {
+        throw ss::httpd::bad_param_exception("group_id cannot be empty");
```
Can this actually happen or it is guaranteed by the router that this is not empty? nit: Anyway, the right error here would be 404.

nvartolomei commented on tests/rptest/services/admin.py:
```diff
@@ -1849,3 +1849,15 @@ def get_debug_bundle_file(self, filename: str, node: MaybeNode = None):
         path = f"debug/bundle/file/{filename}"
         return self._request("DELETE", path, node=node)
+
+    def unsafe_abort_group_transaction(self, group_id: str, pid: int,
```
nit: `    def unsafe_abort_group_transaction(self, group_id: str, *, pid: int,` to force the caller to specify the param names. too many ints that are easy to reorder

nvartolomei commented on src/v/kafka/server/group_data_parser.h:
```diff
@@ -101,17 +101,68 @@ class group_data_parser {
+              dedicated fence batch type (group_tx_fence) was used for group
+              transaction fencing.
+
+              After this buggy compaction, these uncleaned tx_fence batches are
```
Is the buggy compaction fixed now? Otherwise. shouldn't we just fix the buggy compaction instead?

bharathv replied:
we can't fix that easily without a major refactor of the code and we don't need to (as outlined in the comment) as we already moved away from that problematic tx_fence batch infavor of group_tx_fence, so the whole fix revolves around how to get rid of the outdated tx_fence batches.

nvartolomei commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -118,14 +118,9 @@ ss::future<> group_tx_tracker_stm::handle_raft_data(model::record_batch) {
-      model::producer_identity{header.producer_id, header.producer_epoch},
-      header.base_offset);
+  model::record_batch_header, kafka::group_tx::offsets_metadata) {
+    // Transaction boundaries are determined by fence/commit or abort
```
This change doesn't seem to be explained by the commit message?

bharathv replied:
Is split this into a separate commit (and added more changes to it)

nvartolomei reviewed (commented):
Skimmed the PR to be up-to-date with the changes. Hope the observations are useful.

ztlpn commented on src/v/redpanda/admin/api-doc/transaction.json:
```diff
@@ -92,6 +92,46 @@
             ]
+        },
+        {
+            "path": "/v1/transaction/{group_id}/unsafe_abort_group_transaction",
```
Nit: this looks a bit inconsistent, because group_id can be easily confused with transactional_id (when comparing with other endpoints)

bharathv replied:
Thats a good point, I moved the {group_id} to the end of the path.

ztlpn commented on src/v/redpanda/admin/debug.cc:
```diff
@@ -810,6 +817,79 @@ admin_server::get_partition_state_handler(
+      cluster::get_producers_request{ntp, timeout});
+    if (result.error_code != cluster::tx::errc::none) {
+        throw ss::httpd::server_error_exception(fmt::format(
+          "Error {} processing partition state for ntp: {}",
```
nit: "processing partition state" sounds a bit ambiguous, maybe "getting producers for ntp:" instead?

ztlpn commented on src/v/redpanda/admin/debug.cc:
```diff
@@ -810,6 +817,79 @@ admin_server::get_partition_state_handler(
+    const model::ntp ntp = parse_ntp_from_request(req->param);
+    auto timeout = std::chrono::duration_cast<model::timeout_clock::duration>(
+      10s);
+    auto result = co_await _tx_gateway_frontend.local().get_producers(
```
should we check here if the frontend is initialized (as is done for the other endpoint)?

bharathv replied:
I don't think we need to because it is started before creating server, I added just in case.

ztlpn commented on src/v/redpanda/admin/transaction.cc:
```diff
@@ -213,3 +221,73 @@ admin_server::delete_partition_handler(std::unique_ptr<ss::http::request> req) {
+        pid = model::producer_id{parsed_pid};
+    } catch (const boost::bad_lexical_cast& e) {
+        throw ss::httpd::bad_param_exception(
+          fmt::format("invalid producer_id, should be >= 0: {}", e));
```
nit: I guess printing values should be more useful than printing the exception?

ztlpn commented on src/v/redpanda/admin/transaction.cc:
```diff
@@ -213,3 +221,73 @@ admin_server::delete_partition_handler(std::unique_ptr<ss::http::request> req) {
+    auto group_ntp = mapper.ntp_for(kafka::group_id{group_id});
+    if (!group_ntp) {
+        throw ss::httpd::server_error_exception(
+          "consumer_offsets topic now found");
```
typo now -> not

ztlpn commented on src/v/redpanda/admin/debug.cc:
```diff
@@ -810,6 +817,79 @@ admin_server::get_partition_state_handler(
+        producers.producers.push(std::move(producer_state));
+        co_await ss::coroutine::maybe_yield();
+    }
+    co_return ss::json::json_return_type(std::move(producers));
```
`ss::json::stream_range_as_array`?

bharathv replied:
I wanted to do it originally but I didn't because the final output is not array, It is something like
```
{
  ntp = foo,
  producers = [...]
}
```
it there a better way to do this?

ztlpn commented on src/v/cluster/producer_state.h:
```diff
@@ -118,17 +130,19 @@ class requests {
+        return _inflight_requests;
+    }
+
+    const request_queue& fnished_requests() const { return _finished_requests; }
```
typo: fnished -> finished

bharathv replied:
one typo and IDE replicates it everywhere. 🙈

bharathv committed:
raft/persisted_stm: add ability for stms to reject local snapshot

.. in this case the state machine proceeds on to applying from the log.

bharathv committed:
tx/group_tx_stm: invalidate existing snapshots for the stm

Bumps the supported snapshot version so the existing snapshots are
invalidated as they may contain stale max_collectible_offset. This forces
the stm to reconstruct the state form the log and recompute correct
max_collectible_offset.

ztlpn commented on src/v/cluster/tx_protocol_types.h:
```diff
@@ -942,4 +943,120 @@ struct find_coordinator_request
+      , timeout{timeout} {}
+
+    model::ntp ntp;
+    model::timeout_clock::duration timeout;
```
nit: should timeout be in this struct if it is not serialized?

ztlpn replied:
ok now it is serialized

ztlpn commented on tests/rptest/transactions/verifiers/consumer_offsets_verifier.py:
```diff
@@ -0,0 +1,195 @@
+        i = 0
+        while not self._stop_ev.isSet():
+            producer.begin_transaction()
+            producer.produce(self._topic, f"{tx_id}_{i}", f"{tx_id}_id")
```
nit: is it true that in a real application we would actually read from the "source topic" and here we are just simulating it by randomly advancing offsets? IMO this deserves a comment, also maybe move "producing" to happen after "consuming"

bharathv replied:
yes exactly, I think the for purpose of this test a real consumer is just noise, hence generating dummy offsets for validation. will update the comment.

ztlpn commented on tests/rptest/transactions/consumer_offsets_test.py:
```diff
@@ -0,0 +1,119 @@
+        return self.rp_install_version(num_previous=num_previous - 1,
+                                       version=previous)
+
+    def previous_previous_version(self):
```
unused?

ztlpn commented on tests/rptest/transactions/consumer_offsets_test.py:
```diff
@@ -0,0 +1,119 @@
+
+        # At the time of writing, this test checks the following version upgrades
+        # Each upgrade also rolls logs ensures it is compacted
+        # 24.3.x (initial_version) -> 24.4.x
```
there won't be 24.4, only 25.1 :)

bharathv replied:
yikes, good catch.

ztlpn commented on tests/rptest/transactions/consumer_offsets_test.py:
```diff
@@ -0,0 +1,119 @@
+        for v in self.upgrade_through_versions(versions_in=versions,
+                                               already_running=True):
+            self.logger.info(f"Updated to {v}")
+            verifier.verify()
```
is verifier reusable or do we have to recreate it?

bharathv replied:
it is reusable.

bharathv committed:
tx/group: support describe_producers for group

bharathv committed:
group_tx_tracker/stm: plumb feature_table into the stm

bharathv committed:
tx/group/group_tx_tracker: tighten up computation of collectible_offsset

Consider group_metadata to determine if a group transaction should be
considered open. Eg: if a group if tombstoned, any transaction
corresponding to the group is ignored. This invariant is also held in
the actual group stm to ensure groups are not tombstoned before any
pending transactions are cleaned up

bharathv reviewed (commented):
please don't review this just yet, let me do a few ci runs since there are a bunch of changes since last round. (just publishing comments)

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60103-->
<!--387c66a7-9269-4242-83c4-e169d3bd43d6-->
<!--6481ea9b-bfaf-4af2-89ca-7e84e6f2ddd9-->
<!--title-->
## Retry command for [Build#60103](https://buildkite.com/redpanda/redpanda/builds/60103)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/transactions/producers_api_test.py::ProducersAdminAPITest.test_producers_state_api_during_load
```
<!--/body-->

bharathv reviewed (commented):
Ready for review again, main (logical) changes in the latest revision

* transaction tracking stm now uses group metadata to determine if a group is active/can be tombstoned (along with it's transactions).
* persisted stm can now reject snapshots if it wishes to, this is done to invalidate stale snapshots to recompute max_collectible_offset again from log contents (after upgrade).

bharathv commented:
/ci-repeat 5

bharathv commented on tests/rptest/scale_tests/tiered_storage_cache_stress_test.py:
```diff
@@ -65,6 +65,10 @@ def setUp(self):
             self.manifest_upload_interval,
             'disable_public_metrics': False,
             'cloud_storage_cache_check_interval': 500,
+            # We're consuming from leader node and expect specific changes in
```
oops bad rebase, ignore this commit, will fix before merge.

ztlpn commented on src/v/kafka/server/group_data_parser.h:
```diff
@@ -101,17 +101,68 @@ class group_data_parser {
-        if (
-          b.header().type == model::record_batch_type::tx_fence
-          || b.header().type == model::record_batch_type::group_fence_tx) {
+        if (b.header().type == model::record_batch_type::tx_fence) {
```
Should we gate this by the `group_tx_fence_dedicated_batch_type` flag?

ztlpn commented on src/v/raft/persisted_stm.cc:
```diff
@@ -545,17 +545,26 @@ ss::future<> persisted_stm_base<BaseT, T>::start() {
+                  next_offset);
+                _last_snapshot_offset = snapshot.header.offset;
+            } else {
+                next_offset = _raft->start_offset();
```
I wonder if we should leave the next offset as is here. If we set it to the start offset, we will skip application of the raft snapshot, which is not what we want.

bharathv replied:
good catch, done.

ztlpn commented on src/v/raft/persisted_stm.h:
```diff
@@ -223,7 +226,8 @@ class persisted_stm_base
      * Called when local snapshot is applied to the state machine
      */
-    virtual ss::future<> apply_local_snapshot(stm_snapshot_header, iobuf&&) = 0;
+    virtual ss::future<local_snapshot_applied>
```
Nit: would benefit from a comment on how local_snapshot_applied::no is handled

bharathv reviewed (commented):
next push also adds a limit on number of producers fetched via debug output, as a guard rail.

mmaslankaprv commented on src/v/kafka/server/group.h:
```diff
@@ -177,6 +180,7 @@ class group final : public ss::enable_lw_shared_from_this<group> {
         model::timeout_clock::time_point last_update;
 
         bool is_expiration_requested{false};
+        model::offset begin_offset{-1};
```
nit: why we can not use the default value for this offset ?

bharathv replied:
we can, I used it because kafka protocol uses -1 as no offset and its easier for format in the output.

bharathv committed:
k/group_data_parser: reduce ignored batch logging to debug

bharathv committed:
group_tx_tracker/stm: holdoff stm updates during 24.2.x upgrade

This somewhat simplifies cases where the stm needs to disambiguate between
legit tx_fence batches and leftover tx_fence batches from compaction.
(see next commit)

ztlpn commented on src/v/cluster/tx_protocol_types.h:
```diff
@@ -942,4 +943,134 @@ struct find_coordinator_request
+    // Not all producers are sent as a guard rail in extreme cases
+    // with too many producers, a separate count is sent that
+    // denotes the actual number while the vector is a subset.
+    size_t producer_count;
```
let's zero-initialize as a precaution (for the cases with non-zero error code)

ztlpn replied:
Hmm I don't see it fixed?

bharathv replied:
done.

bharathv committed:
group_tx_tracker/stm: track additional information about fence batches

ztlpn commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1620,6 +1621,64 @@ group_manager::describe_group(const model::ntp& ntp, const kafka::group_id& g) {
+      _groups.begin(),
+      _groups.end(),
+      std::back_inserter(groups),
+      [&ntp](auto g_pair) {
```
nit: `const auto&`?

ztlpn commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1739,4 +1798,96 @@ error_code group_manager::validate_group_status(
+          return group->partition()->ntp() == ntp;
+      });
+    reply.producer_count = std::accumulate(
+      _groups.begin(),
+      _groups.end(),
```
should this be `groups` (i.e. the already filtered collection?)

bharathv replied:
good catch, yes.

ztlpn commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1739,4 +1798,96 @@ error_code group_manager::validate_group_status(
+          return acc + entry.second->producers().size();
+      });
+    for (auto& [gid, group] : groups) {
+        if (reply.producers.size() > request.max_producers_to_include) {
```
`>=`?

ztlpn commented on src/v/kafka/server/group_manager.cc:
```diff
@@ -1739,4 +1798,96 @@ error_code group_manager::validate_group_status(
+                producer_info.coordinator_partition = tx->coordinator_partition;
+            }
+            reply.producers.push_back(std::move(producer_info));
+            if (reply.producers.size() > request.max_producers_to_include) {
```
analogously, strictly speaking, shouldn't the condition be `>=` and be checked at the beginning of the loop body?

ztlpn commented on tests/rptest/transactions/transactions_test.py:
```diff
@@ -294,6 +294,45 @@ def simple_test(self):
+
+        producer = ck.Producer({
+            'bootstrap.servers': self.redpanda.brokers(),
+            'transactional.id': '0',
```
nit: maybe some more greppable tx id?

bharathv committed:
group_tx_tracker/stm: heuristic to ignore stale tx_fence batches

bharathv committed:
group_recovery_consumer/logging: tidy up logging

bharathv committed:
group_recovery/tx: fix group recovery for non existent groups

If a group got tombstoned all the producers to that group should be
ignored. The current logic is incorrectly recovering producers and
loading them up to expire later.

bharathv committed:
group_tracker/stm: add a periodic GC loop to expire stale tx_fence txes

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60249-->
<!--19e8f52f-ab96-4fea-85df-15579ea175aa-->
<!--a006d3d3-9336-4a65-ad6f-3157113b5b24-->
<!--title-->
## Retry command for [Build#60249](https://buildkite.com/redpanda/redpanda/builds/60249)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/transactions/consumer_offsets_test.py::VerifyConsumerOffsetsThruUpgrades.test_consumer_group_offsets@{"versions_to_upgrade":3}
```
<!--/body-->

ztlpn commented on src/v/kafka/server/group_tx_tracker_stm.h:
```diff
@@ -72,51 +120,50 @@ class group_tx_tracker_stm final
 class group_tx_tracker_stm_factory : public cluster::state_machine_factory {
 public:
-    group_tx_tracker_stm_factory() = default;
+    group_tx_tracker_stm_factory(ss::sharded<features::feature_table>&);
```
nit: explicit

ztlpn commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -166,18 +272,85 @@ bool group_tx_tracker_stm_factory::is_applicable_for(
+        }
         begin_offsets.emplace(offset);
-        producer_to_begin[pid] = offset;
+        producer_states[pid] = p_state;
```
should we update `producer_to_begin_deprecated` here and elsewhere as well (for the case of a minor downgrade?)

ztlpn commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -166,18 +272,85 @@ bool group_tx_tracker_stm_factory::is_applicable_for(
+          .begin_offset = offset,
+          .batch_ts = begin_ts,
+          .timeout = tx_timeout};
+        if (p_state.expired_deprecated_fence_tx()) {
```
Hmm this makes me a bit uncomfortable as this is non-deterministic... But maybe it's okay because these producer states are not expected to be updated ever.

bharathv replied:
there is a periodic GC loop added in the next commit that sweeps through this list and expires any old stuff

ztlpn commented on src/v/kafka/server/group_tx_tracker_stm.cc:
```diff
@@ -166,18 +272,85 @@ bool group_tx_tracker_stm_factory::is_applicable_for(
+        if (p_state.expired_deprecated_fence_tx()) {
+            vlog(
+              klog.debug,
+              "Ignoring stale tx_fence batch at offset: {}, considering it "
```
nit: some more context? (e.g. group id, producer id)

vbotbuildovich commented:
/backport v24.3.x

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
Failed to create a backport PR to v24.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24637-v24.3.x-292 remotes/upstream/v24.3.x
git cherry-pick -x 89478480b4 cdd274a2a5 24c8e89f4f 8c5ecca7b5 3736f00b6a 835f3fce5d 9eee6326c5 f7191ad708 6ca81b456f 3d65c3933d 6fc62bb05b 2b79687b9a ac2204129a 7c8d6335a3 c7f953efd9 9958ca60a0 6efd325f85 23c8e295a7 70e36ebe57 c833f50286 0051463f23
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12603866076).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-24637-v24.2.x-818 remotes/upstream/v24.2.x
git cherry-pick -x 89478480b4 cdd274a2a5 24c8e89f4f 8c5ecca7b5 3736f00b6a 835f3fce5d 9eee6326c5 f7191ad708 6ca81b456f 3d65c3933d 6fc62bb05b 2b79687b9a ac2204129a 7c8d6335a3 c7f953efd9 9958ca60a0 6efd325f85 23c8e295a7 70e36ebe57 c833f50286 0051463f23
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/12603866450).
